### ISSUE TITLE
Extensibility point for loggers

### ISSUE THREAD

**Issue by vivek1729** [nteract/archived-desktop-app#5260]
Hi everyone,

I am wondering if there is a general logger interface available in Nteract. For instance, if users are consuming an nteract package in their larger application, they might want to log some information from within the nteract package (if the package supports it) for their own telemetry and feedback loop.
I am working on adding support for the monaco-editor package, and I wanted to know if there's any general guidance on adding this support. I am not sure if there is a central logging solution in Nteract that enables this.

For eg: In any package, we could have a prop that takes in a logger with the following generic interface:

```
/**
 * Logger
 */
export interface ILogger {
  error: (msg: any) => void;
  warn: (msg: any) => void;
  info: (msg: any) => void;
  log: (msg: any) => void;
}
```

**Comment by captainsafia** [comment]
Great question, @vivek1729!

nteract really only provides "logging" support in one place at the moment, we have a Redux middleware logger that logs actions and state to the console.

With regard to a general logging interface, there isn't really something specific and opinionated in nteract. I'd hesitate to introduce a logging abstraction within nteract itself. Instead, I'd be curious to see what best practices are recommended for logging in React/Redux and make sure our design lends itself to those best practices, for example, registering a middleware to support logging in Redux (which can happen completely independently of nteract).

Your comment about the monaco-editor brings up the question of logging in React components. For some React components, specifically outputs, we leverage [error boundaries](https://reactjs.org/docs/error-boundaries.html) to capture errors during rendering.

If we want to pursue this model further, we can make sure to implement error boundaries in each component and allow users to provide a `error` method that gets invoked inside `componentDidCatch`. Note, this model doesn't provide support for other log levels, like `warn` and `info`. 

TL;DR: No nteract-specific logging infrastructure. Since nteract is built on top of React + Redux, users can leverage logging middlewares in Redux and error boundaries in React components.

Logging is one of those things that can quickly get out of hand so I'm cautious about building something bespoke and would like to rely on industry best practices.

@nathanieldeal @emeeks @MSeal @willingc @CrystallineCat  Thoughts on this?

**Comment by vivek1729** [comment]
Thanks for the prompt feedback @captainsafia . Error boundaries is an interesting idea that I wasn't familiar with. What I wanted to understand better was just the support for a logger inside react components or nteract packages that people might use in isolation as well. As you rightly pointed out, the redux logger and error boundaries are limited in their domain. Instead of having a full fledged logging infrastructure, I was proposing that we introduce a generic logger interface `ILogger` which is fairly  standard as far as the signature of the `error`, `warn` and `info` methods are concerned. Then, we just add an optional `logger` prop to the React components in Nteract. So, any consumer of those nteract packages can hook into that prop and provide their own logger implementation that follows the `ILogger` interface. This would allow consumers to catch all kinds of errors/warnings that might be emitted inside the package.

**Comment by captainsafia** [comment]
>  This would allow consumers to catch all kinds of errors/warnings that might be emitted inside the package.

The "might be emitted inside the package" is the key part here. Regardless of what interface we provide, we'd have to go through each component and add the logging invocations to it. This is ultimately where a bulk of the work and maintenance cost will come from. The subtleties here include:

* we need to determine what scenarios we want to log for
* we need to determine what log levels those scenarios should be logged at
* once implemented, we'll need to account for new logging requests from end users ("hey, can we add a log.info in this component")

Adding the ILogger interface is straightforward enough. Going in and adding the actual logging statements to the components will require a lot more effort. Do you have thoughts on how we can manage this effort?

> Then, we just add an optional logger prop to the React components in Nteract.

I'd consider maybe doing this as an epic that processes `LOG` actions dispatched from the component. This would centralize the logic for logging a little bit. We can implement the ILogger interface as an object that is injected into the Epic middleware as a dependency (similar to what we do with the ContentProviders today).

**Comment by vivek1729** [comment]
Thanks for the feedback @captainsafia. Trying to understand your proposal a little better. You're suggesting the users could emit `LOG` actions for any errors they want to send across. So from a consumption point of you, any user wanted to log errors emitted from the components can define their own epic to do so. Is my understanding correct?

I understand your point around adding the support to the component. Just to clarify a little bit, my suggestion was not to have users emit logs for scenarios specific to their own needs. It was centered more around `tracing` and being able to surface the errors, exceptions and warnings in different nteract components to the end user.

**Comment by willingc** [comment]
> It was centered more around tracing and being able to surface the errors, exceptions and warnings in different nteract components to the end user.

@vivek1729 Thanks for the additional context. Question about the above statement: are you looking at this from the standpoint of a developer extending nteract or a user of nteract itself?

**Comment by vivek1729** [comment]
@willingc thanks for that clarifying question.  When I say user, I didn't imply someone using the nteract desktop app or jupyter extension. Rather, I was alluding to a developer extending nteract like you pointed out. For instance, someone relying on nteract's editor components as part of their larger project. In that scenario, they might want to capture and act upon errors that stem from the nteract components

**Comment by captainsafia** [comment]
> So from a consumption point of you, any user wanted to log errors emitted from the components can define their own epic to do so. Is my understanding correct?

Correct. They could use their own epic or the one we bundle. A scenario where they might use their own epic is if they wanted to aggregate and dispatch their logs to a monitoring service like [Sentry](https://sentry.io/welcome/).

> It was centered more around tracing and being able to surface the errors, exceptions and warnings in different nteract components to the end user.

Yep. I understood that. My point was that we needed to come up with a plan on what needs to be logged in each component since it is the SDK's responsibility to emit the LOG actions (based on this proposal).

**Comment by vivek1729** [comment]
Got it. Thanks for clarifying that @captainsafia. I am not really sure what the ideal plan should be. It seems like the work needed to emit errors as log actions would follow a single pattern where we either pass the action as a prop to stateless components or dispatch it directly from the stateful ones. We can consider taking an incremental approach where let's say we do it for one stateful component and it's child components, establish the pattern, get feedback and then do it for the other components.

It will be interesting to draw from a prior exercise of this sort.
Curious what strategy have we followed in the past.

**Comment by captainsafia** [comment]
Woop! Had a comment written here that I never posted.

An incremental approach sounds good. I would take a look at the way things evolved in the RFC to move stuff out of the monorepo (#4981). An ordered list of packages to add logs to should suffice.

Starting off with stateful-components sounds like a good idea. I'd probably put outputs and up high on the list too (particular jupyter-widgets).
